sqlglot/optimizer/qualify_columns.py

```diff
@@ -380,18 +410,89 @@ def _qualify_columns(scope: Scope, resolver: Resolver) -> None:
                     column.set("table", column_table)
 
 
+def _expand_struct_stars(
+    expression: exp.Dot,
+    next_alias_name: t.Callable[[], str],
+) -> t.List[exp.Alias]:
+    struct_fields: t.List[str] = []
+
+    # [BigQuery] Expand/Flatten foo.bar.* where bar is a struct column
+    starting_col = expression.find(exp.Column)
+    assert starting_col
+
+    dot_parts = expression.parts
+    table_name = starting_col.table
+
+    # If there's more nested levels specified, start the DFS traversal from the nested struct first
+    for part in dot_parts[2:-1]:
+        datatype = starting_col.find(exp.DataType) or (
+            starting_col.type and starting_col.type.find(exp.DataType)
+        )
+        assert datatype
```

Comment:
This is to find _any_ data type even if it's `Type.Unknown`, so we always expect to have a valid `datatype` afaict. All the asserts are to eliminate mypy `t.Optional` warnings without checking them with branches, as there would be no reason to have `None` expressions throughout the `expand_struct_star`

Reply:
I think this is risky. We should _only_ assert when we're 100% sure about a runtime property and we *need* it to hold. The logic here is somewhat implicit and feels tricky to debug.

For example, if you get an `UNKNOWN` type from the type inference, you'll silently skip the loop below because `datatype.expressions` will return an empty list for it. And then if you hit L441, 1) you'll have a `starting_col` and 2) it looks like the `starting_col.this.this != dot_parts[-2].this` condition will fail? Hence we'll raise.

Let's make sure this is implemented in a more explicit fashion to safely handle all code paths and avoid raising.